components/places/src/import/ios_bookmarks.rs

```diff
@@ -0,0 +1,566 @@
+    conn.execute_batch(&POPULATE_STAGING)?;
+    scope.err_if_interrupted()?;
+
+    log::debug!("Populating missing entries in moz_places");
+    conn.execute_batch(&FILL_MOZ_PLACES)?;
+    scope.err_if_interrupted()?;
+
+    log::debug!("Populating mirror");
+    conn.execute_batch(&POPULATE_MIRROR)?;
+    scope.err_if_interrupted()?;
+
+    log::debug!("Populating mirror tags");
+    populate_mirror_tags(&conn)?;
+    scope.err_if_interrupted()?;
+
+    // Ideally we could just do this right after `CREATE_AND_POPULATE_STAGING`,
+    // which would mean we could detach the iOS database sooner, but we have
+    // constraints on the mirror structure that prevent this (and there's
+    // probably nothing bad that can happen in this case anyway).
+    log::debug!("Populating mirror structure");
+    conn.execute_batch(&POPULATE_MIRROR_STRUCTURE)?;
+    scope.err_if_interrupted()?;
+
+    // log::debug!("Detaching iOS database");
+    // drop(auto_detach);
+    // scope.err_if_interrupted()?;
+
+    let store = BookmarksStore::new(&conn, &scope);
+    let mut merger = Merger::new(&store, Default::default());
+    // We're already in a transaction.
+    merger.set_external_transaction(true);
+    log::debug!("Merging with local records");
+    merger.merge()?;
+    scope.err_if_interrupted()?;
+
+    // Update last modification time, sync status, etc
+    log::debug!("Fixing up bookmarks");
+    conn.execute_batch(&FIXUP_MOZ_BOOKMARKS)?;
+    scope.err_if_interrupted()?;
+    log::debug!("Committing...");
+    tx.commit()?;
+
+    // Note: update_frecencies manages it's own transaction, which is fine,
+    // since nothing that bad will happen if it is aborted.
+    log::debug!("Updating frecencies");
+    store.update_frecencies()?;
+
+    log::info!("Successfully imported bookmarks!");
+
+    auto_detach.execute_now()?;
+    clear_mirror_on_drop.execute_now()?;
+
+    Ok(())
+}
+
+// If we must.
+fn populate_mirror_tags(db: &crate::PlacesDb) -> Result<()> {
+    use crate::storage::tags::{validate_tag, ValidatedTag};
+    let mut tag_map: HashMap<String, Vec<i64>> = HashMap::new();
+    {
+        let mut stmt = db.prepare(
+            "SELECT mirror.id, stage.tags
+             FROM main.moz_bookmarks_synced mirror
+             JOIN temp.iosBookmarksStaging stage USING(guid)
+             -- iOS tags are JSON arrays of strings (or null).
+             -- Both [] and null are allowed for 'no tags'
+             WHERE stage.tags IS NOT NULL
+               AND stage.tags != '[]'",
+        )?;
+
+        let mut rows = stmt.query(NO_PARAMS)?;
+        while let Some(row) = rows.next()? {
+            let id: i64 = row.get(0)?;
+            let tags: String = row.get(1)?;
+            let tag_vec = if let Ok(ts) = serde_json::from_str::<Vec<String>>(&tags) {
+                ts
+            } else {
+                log::warn!("Ignoring bad `tags` entry");
+                log::warn!("  entry had {:?}", tags);
```

Comment:
Trace for PII?

Reply:
Nice catch, it was separate so that we could trace it, and then... I guess I forgot.